libs/@blockprotocol/graph/src/shared.ts

```diff
@@ -0,0 +1,28 @@
+type TupleEntry<
+  T extends readonly unknown[],
+  I extends unknown[] = [],
+  R = never,
+> = T extends readonly [infer Head, ...infer Tail]
+  ? TupleEntry<Tail, [...I, unknown], R | [`${I["length"]}`, Head]>
+  : R;
+type ObjectEntry<T extends {}> = T extends object
```

Comment:
```suggestion

type ObjectEntry<T extends {}> = T extends object
```

Reply:
I left these grouped together to try and encourage some intuition that they're all directly related, but most importantly they're implementation details for the thing below (which has a comment sharing the source).

I'm happy to add the newlines if you'd prefer just for general style consistency though.

Reply:
I think it makes it easier to read, yes. We don't do this grouping elsewhere for types which are just utilities for types below them